File: agentcore-data-governance/models/data_elements.py
"""
Data element models for the Agentic Data Governance System.

This module defines Pydantic models for data elements, mappings, and gaps.

Requirements: 2.2
"""

from datetime import datetime
from typing import Literal, Optional
from pydantic import BaseModel, Field
from uuid import uuid4


# Type aliases
DataType = Literal['string', 'number', 'date', 'boolean', 'decimal', 'integer']
DataGapReason = Literal['no_source', 'partial_source', 'calculation_needed']
ArtifactStatus = Literal['draft', 'pending_review', 'approved', 'rejected']
ReconciliationItemStatus = Literal['matched', 'added', 'removed', 'modified']


class DataElement(BaseModel):
    """Represents a data element from a regulatory template."""
    id: str = Field(default_factory=lambda: str(uuid4()))
    name: str
    regulatory_definition: str
    data_type: DataType
    format: str
    calculation_logic: Optional[str] = None
    unit: Optional[str] = None
    mandatory: bool


class DataMapping(BaseModel):
    """Mapping of a data element to internal sources."""
    element_id: str
    source_system: str
    source_table: str
    source_field: str
    transformation_logic: Optional[str] = None
    confidence: float
    validated_by: Optional[str] = None


class DataGap(BaseModel):
    """Represents a data gap where no internal source is found."""
    element_id: str
    element_name: str
    reason: DataGapReason
    suggested_resolution: Optional[str] = None


class RequirementsDocument(BaseModel):
    """Requirements document for a regulatory report."""
    id: str = Field(default_factory=lambda: str(uuid4()))
    report_id: str
    elements: list[DataElement] = []
    mappings: list[DataMapping] = []
    gaps: list[DataGap] = []
    version: int = 0
    status: ArtifactStatus = 'draft'
    created_at: datetime
    updated_at: datetime
    validated_by: Optional[str] = None
    validated_at: Optional[datetime] = None


class ReconciliationItem(BaseModel):
    """Individual item in reconciliation result."""
    item_id: str
    item_type: str
    status: ReconciliationItemStatus
    existing_value: Optional[dict] = None
    new_value: Optional[dict] = None
    differences: list[str] = []


class ReconciliationResult(BaseModel):
    """Result of reconciling documents."""
    items: list[ReconciliationItem] = []
    matched_count: int = 0
    added_count: int = 0
    removed_count: int = 0
    modified_count: int = 0

File: src/types/data-elements.ts
/**
 * Data element and requirements types for the Agentic Data Governance System
 */

import { DataType, DataGapReason, ArtifactStatus, ReconciliationItemStatus } from './common.js';

/**
 * Represents a data element from a regulatory template
 */
export interface DataElement {
  id: string;
  name: string;
  regulatoryDefinition: string;
  dataType: DataType;
  format: string;
  calculationLogic?: string;
  unit?: string;
  mandatory: boolean;
}

/**
 * Mapping of a data element to internal sources
 */
export interface DataMapping {
  elementId: string;
  sourceSystem: string;
  sourceTable: string;
  sourceField: string;
  transformationLogic?: string;
  confidence: number;
  validatedBy?: string;
}

/**
 * Represents a data gap where no internal source is found
 */
export interface DataGap {
  elementId: string;
  elementName: string;
  reason: DataGapReason;
  suggestedResolution?: string;
}

/**
 * Requirements document for a regulatory report
 */
export interface RequirementsDocument {
  id: string;
  reportId: string;
  elements: DataElement[];
  mappings: DataMapping[];
  gaps: DataGap[];
  version: number;
  status: ArtifactStatus;
  createdAt: Date;
  updatedAt: Date;
  validatedBy?: string;
  validatedAt?: Date;
}

/**
 * Result of reconciling documents
 */
export interface ReconciliationResult {
  items: ReconciliationItem[];
  matchedCount: number;
  addedCount: number;
  removedCount: number;
  modifiedCount: number;
}

/**
 * Individual item in reconciliation result
 */
export interface ReconciliationItem {
  itemId: string;
  itemType: string;
  status: ReconciliationItemStatus;
  existingValue?: unknown;
  newValue?: unknown;
  differences?: string[];
}

/**
 * Data catalog for internal sources
 */
export interface DataCatalog {
  id: string;
  systems: DataSystem[];
  lastUpdated: Date;
}

/**
 * Represents a data system in the catalog
 */
export interface DataSystem {
  id: string;
  name: string;
  tables: DataTable[];
}

/**
 * Represents a table in a data system
 */
export interface DataTable {
  id: string;
  name: string;
  fields: DataField[];
}

/**
 * Represents a field in a data table
 */
export interface DataField {
  id: string;
  name: string;
  dataType: DataType;
  description?: string;
  businessTerm?: string;
}
